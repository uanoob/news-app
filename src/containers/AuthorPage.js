import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import Avatar from '@material-ui/core/Avatar';
import IconButton from '@material-ui/core/IconButton';
import AddIcon from '@material-ui/icons/AddCircleOutline';
import Articles from '../components/article/Articles';
import ArticleForm from '../components/article/ArticleForm';
import { handleAuthorAvatar, stringToColor } from '../utils';

const styles = theme => ({
  container: {
    margin: 'auto',
    marginTop: 20,
    width: '100%',
    maxWidth: 600,
    backgroundColor: theme.palette.background.paper,
  },
});

class AuthorPage extends Component {
  state = {
    articleDialog: false,
  };

  handleArticleClick = () => {
    const { articleDialog } = this.state;
    this.setState({
      articleDialog: !articleDialog,
    });
  };

  render() {
    const {
      classes, match, userId, userName, userEmail,
    } = this.props;
    const { articleDialog } = this.state;
    return (
      <div>
        <div className={classes.container}>
          {match.params.id === userId ? (
            <Card>
              <CardHeader
                avatar={(
                  <Avatar
                    aria-label="Recipe"
                    style={{
                      backgroundColor: stringToColor(userName || 'Anonymous'),
                    }}
                  >
                    {handleAuthorAvatar(userName || 'Anonymous')}
                  </Avatar>
)}
                action={(
                  <IconButton onClick={this.handleArticleClick}>
                    <AddIcon />
                  </IconButton>
)}
                title={userName}
                subheader={userEmail}
              />
            </Card>
          ) : null}
          <Articles authorId={match.params.id} />

          <ArticleForm
            dialog={articleDialog}
            handleDialogClick={this.handleArticleClick}
            header="Add an article"
            description="Please add your article here."
          />
        </div>
      </div>
    );
  }
}
AuthorPage.propTypes = {
  classes: PropTypes.shape({
    container: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    isExact: PropTypes.bool.isRequired,
    params: PropTypes.shape({ id: PropTypes.string.isRequired }).isRequired,
    path: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  }).isRequired,
  userId: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  userEmail: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  userId: state.user.user._id,
  userName: state.user.user.name,
  userEmail: state.user.user.email,
});

const mapDispatchToProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(withStyles(styles)(AuthorPage));
