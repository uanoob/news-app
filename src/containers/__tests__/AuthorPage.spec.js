import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import Card from '@material-ui/core/Card';
import AuthorPage, { AuthorPageComponent } from '../AuthorPage';

describe('AuthorPageComponent', () => {
  const initialState = {
    articleDialog: false,
  };
  const props = {
    classes: {
      container: '',
    },
    match: {
      isExact: true,
      path: 'test-path',
      url: 'test-url',
      params: {
        id: 'test-is-author-id',
      },
    },
    userId: 'test-is-author-id',
    userName: 'test-userName',
    userEmail: 'test-userEmail',
  };

  describe('initialize AuthorPageComponent', () => {
    it('render correctly AuthorPageComponent component', () => {
      const wrapper = shallow(<AuthorPageComponent {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
    it('initialize AuthorPageComponent with initial state', () => {
      const wrapper = shallow(<AuthorPageComponent {...props} />);
      expect(wrapper.state()).toEqual(initialState);
    });
  });
  describe('render <Card /> if user is author', () => {
    it('render <Card />', () => {
      const wrapper = shallow(<AuthorPageComponent {...props} />);
      expect(wrapper.find(Card)).toHaveLength(1);
    });
  });
  describe('not render <Card /> if user not author', () => {
    const nextProps = {
      ...props,
      match: {
        isExact: true,
        path: 'test-path',
        url: 'test-url',
        params: {
          id: 'test-not-author-id',
        },
      },
    };
    const wrapper = shallow(<AuthorPageComponent {...nextProps} />);
    it('not render <Card />', () => {
      expect(wrapper.find(Card)).toHaveLength(0);
    });
  });
  describe('AuthorPage: mapStateToProps', () => {
    const mockStore = configureMockStore();
    const initialStoreState = {
      user: {
        user: {
          _id: 'test-user-id',
        },
      },
    };
    let store;
    beforeEach(() => {
      store = mockStore(initialStoreState);
    });
    it('the state values were correctly passed as props', () => {
      const wrapper = shallow(<AuthorPage {...props} store={store} />);
      expect(wrapper.props().store.getState()).toEqual(initialStoreState);
    });
  });
});
