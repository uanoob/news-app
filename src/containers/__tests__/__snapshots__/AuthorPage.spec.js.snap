// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorPageComponent initialize AuthorPageComponent render correctly AuthorPageComponent component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorPageComponent
    classes={
      Object {
        "container": "",
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {
          "id": "test-is-author-id",
        },
        "path": "test-path",
        "url": "test-url",
      }
    }
    userEmail="test-userEmail"
    userId="test-is-author-id"
    userName="test-userName"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className=""
      >
        <WithStyles(Card)>
          <WithStyles(CardHeader)
            action={
              <WithStyles(IconButton)
                id="article-add-button"
                onClick={[Function]}
              >
                <pure(AddCircleOutlineIcon) />
              </WithStyles(IconButton)>
            }
            avatar={
              <WithStyles(Avatar)
                aria-label="Recipe"
                style={
                  Object {
                    "backgroundColor": "#d13f9b",
                  }
                }
              >
                TE
              </WithStyles(Avatar)>
            }
            subheader="test-userEmail"
            title="test-userName"
          />
        </WithStyles(Card)>
        <Connect(WithStyles(ArticlesComponent))
          authorId="test-is-author-id"
        />
        <Connect(WithStyles(ArticleFormComponent))
          description="Please add your article here."
          dialog={false}
          handleDialogClick={[Function]}
          header="Add an article"
        />
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Card)>
            <WithStyles(CardHeader)
              action={
                <WithStyles(IconButton)
                  id="article-add-button"
                  onClick={[Function]}
                >
                  <pure(AddCircleOutlineIcon) />
                </WithStyles(IconButton)>
              }
              avatar={
                <WithStyles(Avatar)
                  aria-label="Recipe"
                  style={
                    Object {
                      "backgroundColor": "#d13f9b",
                    }
                  }
                >
                  TE
                </WithStyles(Avatar)>
              }
              subheader="test-userEmail"
              title="test-userName"
            />
          </WithStyles(Card)>,
          <Connect(WithStyles(ArticlesComponent))
            authorId="test-is-author-id"
          />,
          <Connect(WithStyles(ArticleFormComponent))
            description="Please add your article here."
            dialog={false}
            handleDialogClick={[Function]}
            header="Add an article"
          />,
        ],
        "className": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(CardHeader)
              action={
                <WithStyles(IconButton)
                  id="article-add-button"
                  onClick={[Function]}
                >
                  <pure(AddCircleOutlineIcon) />
                </WithStyles(IconButton)>
              }
              avatar={
                <WithStyles(Avatar)
                  aria-label="Recipe"
                  style={
                    Object {
                      "backgroundColor": "#d13f9b",
                    }
                  }
                >
                  TE
                </WithStyles(Avatar)>
              }
              subheader="test-userEmail"
              title="test-userName"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "action": <WithStyles(IconButton)
                id="article-add-button"
                onClick={[Function]}
              >
                <pure(AddCircleOutlineIcon) />
              </WithStyles(IconButton)>,
              "avatar": <WithStyles(Avatar)
                aria-label="Recipe"
                style={
                  Object {
                    "backgroundColor": "#d13f9b",
                  }
                }
              >
                TE
              </WithStyles(Avatar)>,
              "subheader": "test-userEmail",
              "title": "test-userName",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "authorId": "test-is-author-id",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "description": "Please add your article here.",
            "dialog": false,
            "handleDialogClick": [Function],
            "header": "Add an article",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className=""
        >
          <WithStyles(Card)>
            <WithStyles(CardHeader)
              action={
                <WithStyles(IconButton)
                  id="article-add-button"
                  onClick={[Function]}
                >
                  <pure(AddCircleOutlineIcon) />
                </WithStyles(IconButton)>
              }
              avatar={
                <WithStyles(Avatar)
                  aria-label="Recipe"
                  style={
                    Object {
                      "backgroundColor": "#d13f9b",
                    }
                  }
                >
                  TE
                </WithStyles(Avatar)>
              }
              subheader="test-userEmail"
              title="test-userName"
            />
          </WithStyles(Card)>
          <Connect(WithStyles(ArticlesComponent))
            authorId="test-is-author-id"
          />
          <Connect(WithStyles(ArticleFormComponent))
            description="Please add your article here."
            dialog={false}
            handleDialogClick={[Function]}
            header="Add an article"
          />
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithStyles(Card)>
              <WithStyles(CardHeader)
                action={
                  <WithStyles(IconButton)
                    id="article-add-button"
                    onClick={[Function]}
                  >
                    <pure(AddCircleOutlineIcon) />
                  </WithStyles(IconButton)>
                }
                avatar={
                  <WithStyles(Avatar)
                    aria-label="Recipe"
                    style={
                      Object {
                        "backgroundColor": "#d13f9b",
                      }
                    }
                  >
                    TE
                  </WithStyles(Avatar)>
                }
                subheader="test-userEmail"
                title="test-userName"
              />
            </WithStyles(Card)>,
            <Connect(WithStyles(ArticlesComponent))
              authorId="test-is-author-id"
            />,
            <Connect(WithStyles(ArticleFormComponent))
              description="Please add your article here."
              dialog={false}
              handleDialogClick={[Function]}
              header="Add an article"
            />,
          ],
          "className": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(CardHeader)
                action={
                  <WithStyles(IconButton)
                    id="article-add-button"
                    onClick={[Function]}
                  >
                    <pure(AddCircleOutlineIcon) />
                  </WithStyles(IconButton)>
                }
                avatar={
                  <WithStyles(Avatar)
                    aria-label="Recipe"
                    style={
                      Object {
                        "backgroundColor": "#d13f9b",
                      }
                    }
                  >
                    TE
                  </WithStyles(Avatar)>
                }
                subheader="test-userEmail"
                title="test-userName"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "action": <WithStyles(IconButton)
                  id="article-add-button"
                  onClick={[Function]}
                >
                  <pure(AddCircleOutlineIcon) />
                </WithStyles(IconButton)>,
                "avatar": <WithStyles(Avatar)
                  aria-label="Recipe"
                  style={
                    Object {
                      "backgroundColor": "#d13f9b",
                    }
                  }
                >
                  TE
                </WithStyles(Avatar)>,
                "subheader": "test-userEmail",
                "title": "test-userName",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "authorId": "test-is-author-id",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "description": "Please add your article here.",
              "dialog": false,
              "handleDialogClick": [Function],
              "header": "Add an article",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
